src/chainstate/stacks/index/marf.rs

```diff
@@ -0,0 +1,2180 @@
+        }
+        Ok(())
+    }
+
+    pub fn insert_leaf(storage: &mut TrieFileStorage, block_hash: &BlockHeaderHash, path: &TriePath, value: &TrieLeaf) -> Result<(), Error> {
+        MARF::do_insert_leaf(storage, block_hash, path, value, true)
+    }
+    
+    // like insert_leaf, but don't update the merkle skiplist
+    pub fn insert_leaf_in_batch(storage: &mut TrieFileStorage, block_hash: &BlockHeaderHash, path: &TriePath, value: &TrieLeaf) -> Result<(), Error> {
+        MARF::do_insert_leaf(storage, block_hash, path, value, false)
+    }
+
+    /// Instantiate the MARF from a TrieFileStorage instance 
+    pub fn from_storage(storage: TrieFileStorage) -> MARF {
+        MARF {
+            storage: storage,
+            open_chain_tip: None,
+        }
+    }
+
+    /// Instantiate the MARF using a TrieFileStorage instance, from the given path on disk.
+    /// This will have the side-effect of instantiating a new fork table from the tries encoded on
+    /// disk. Performant code should call this method sparingly.
+    pub fn from_path(path: &String) -> Result<MARF, Error> {
+        let mut file_storage = TrieFileStorage::new(path)?;
+        match fs::metadata(path) {
+            Ok(_) => {},
+            Err(e) => {
+                if e.kind() != io::ErrorKind::NotFound {
+                    return Err(Error::IOError(e));
+                }
+
+                MARF::format(&mut file_storage, &TrieFileStorage::block_sentinel())?;
+            }
+        };
+
+        Ok(MARF::from_storage(file_storage))
+    }
+
+    /// Resolve a key from the MARF to a MARFValue with respect to the given block height.
+    pub fn get(&mut self, block_hash: &BlockHeaderHash, key: &String) -> Result<Option<MARFValue>, Error> {
+        let cur_block_hash = self.storage.get_cur_block();
+        let cur_block_rw = self.storage.readwrite();
+
+        let path = TriePath::from_key(key);
+        let leaf_opt_res = MARF::get_path(&mut self.storage, block_hash, &path)?;
+
+        // restore
+        self.storage.open_block(&cur_block_hash, cur_block_rw)?;
+
+        match leaf_opt_res {
+            None => {
+                Ok(None)
+            },
+            Some(leaf) => {
+                Ok(Some(leaf.data))
+            }
+        }
+    }
+
+    /// Insert the given (key, value) pair into the MARF.  Inserting the same key twice silently
+    /// overwrites the existing key.  Succeeds if there are no storage errors.
+    /// Must be called after a call to .begin() (will fail otherwise)
+    pub fn insert(&mut self, key: &String, value: &String) -> Result<(), Error> {
+        match self.open_chain_tip {
+            None => {
+                Err(Error::WriteNotBegunError)
+            },
+            Some(ref block_hash) => {
+                let cur_block_hash = self.storage.get_cur_block();
+                let cur_block_rw = self.storage.readwrite();
+
+                let marf_value = MARFValue::from_value(value);
```

Comment:
it wasn't clear on my first read, but the interface for the `insert` functions is pretty difficult to use -- 

The `insert` takes a string `value`, but it doesn't actually store the value -- instead it opaquely hashes it and stores that hash (which `get` would return, _not_ the value itself). This interface would be impossible to use for a caller who is managing side storage for large values themselves (which is probably the best approach -- managing the item storage seems like an orthogonal concern to the marf index).

I think an interface like:

```rust
insert(&mut self, key: &[u8; 32], value: &[u8; 40]) -> Result<(), Error>;
get(&mut self, block_hash: &BlockHeaderHash, key: &[u8; 32]) -> Result<Option<MARFValue>, Error>;
```
Would be more ergonomic.

Reply:
> This interface would be impossible to use for a caller who is managing side storage for large values themselves (which is probably the best approach

I agree that something needs to change, but I don't think it's this.  When you insert data, the algorithm should be:

```
1. Get the MARF hash of the value
2. Store the key/value-hash in the MARF
3. Store the value and value-hash in the data store
```

To read data, the algorithm is:
```
1. Given the chain tip and key, get the MARF value hash
2. Given the value hash, get the associated value
3. (optional) Verify the value hashes to the value hash
```

So, the MARF interface would be:

```rust
fn get_value_hash(value: &String) -> MARFValue;
fn insert(&mut self, key: &String, value_hash: &MARFValue) -> Result<(), Error>;
fn get(&mut self, block_hash: &BlockHeaderHash, key: &String) -> Result<Option<MARFValue>, Error>
```

And, the interface for `MARFValue` would be:

```rust
fn from_value(s: &String) -> MARFValue;
fn verify(&self, s: &String) -> bool;
```

Reply:
These two hashes are only ever computed in those interface functions, so it doesn't seem like something that they should necessarily be required to do -- callers can quite easily perform that themselves.

The other issue for integration (a much larger issue) is that `get`s require a mutable reference. Part of this is because of the usage tracking, but also part of it is because of the structure of the MarfFileStorage struct.

Reply:
> These two hashes are only ever computed in those interface functions, so it doesn't seem like something that they should necessarily be required to do -- callers can quite easily perform that themselves.

Why do you think that the caller should be expected to provide two "bare" byte slices as arguments?  I can't see why the caller should never need to know or care about the `TriePath` structure or the `MARFValue` hash, because presumably the caller only cares about building an index over key/value pairs, and not the particulars of how the index itself is calculated.

> The other issue for integration (a much larger issue) is that gets require a mutable reference. Part of this is because of the usage tracking, but also part of it is because of the structure of the MarfFileStorage struct.

I'm not sure why this is a problem?  Regardless of how the MARF is structure, a mutable reference is always going to be requried as a direct consequence of having to deal with a file descriptor internally.  You _don't want_ concurrent access to it, since it itself has internal mutable state like an offset, a read-ahead buffer, and a last-error.

Reply:
> I can see why you might want to avoid String since we're really dealing with serialized Clarity objects. I'd be happy to take a Vec<u8> instead for each of these, but I can't see how supplying "bare" byte slices with oddly-specific lengths that are dependent on the internal implementation improves things.

Because the MARF doesn't need to be responsible for either hashing keys or managing the actual storage of values.

> I'm not sure why this is a problem?

The borrow checker gets pretty unhappy if every function requires mutable pointers (which would be the case for the Clarity VM to use `.get`)

Reply:
> Because the MARF doesn't need to be responsible for either hashing keys or managing the actual storage of values.

Actually, it does -- the whole point of hashing is to ensure that the keys are evenly distributed, no matter what the schedule of inserts is.  This is particularly important to us, because one attack avenue a network adversary has is to hammer the MARF index by inserting and reading many keys that are close to each other in the key space (since these paths will have the most nodes in them and will thus incur the highest I/O costs to service).  Even if they are suitably paid for, it still slows down the peer.  At some point, I was going to audit the code to ensure that the insert/get paths to include the VRF seed hash in order to ensure that keys are evenly distributed, no matter how the contract data are structured.

> The borrow checker gets pretty unhappy if every function requires mutable pointers (which would be the case for the Clarity VM to use .get)

The only solution I can think of is to refactor `get` to open/close a file handle internally as a stack variable.  We'd want to think very carefully about this because it can lead to a resource exhaustion attack -- i.e. we'll want to limit the number of concurrent calls to `.get()` globally.

Reply:
> Actually, it does -- the whole point of hashing is to ensure that the keys are evenly distributed, no matter what the schedule of inserts is.

Okay, that's fine then.

> The only solution I can think of is to refactor get to open/close a file handle internally as a stack variable. We'd want to think very carefully about this because it can lead to a resource exhaustion attack -- i.e. we'll want to limit the number of concurrent calls to .get() globally.

The mutability won't be too much of an issue for now, since the Clarity VM does a lot more cloning than it really needs to. As we try to eliminate spurious clones, though, the borrow checker's mutability checks will start to scream at us.

However, the file handle issue you mention is already present: any back pointer traversed opens a file descriptor. Making `get` work on an immutable pointer would only really add one more file descriptor open.

Reply:
> However, the file handle issue you mention is already present: any back pointer traversed opens a file descriptor. Making get work on an immutable pointer would only really add one more file descriptor open.

Understood.

The resource exhaustion attack I had in mind becomes relevant once we hook the RPC and P2P network interfaces up to the Clarity VM database to allow clients to query key/value pairs.  Because these calls can be concurrent, the MARF's `get` will either need to take an immtuable self reference (possibly by way of creating the initial file descriptor on the stack), or we'll want to create a bunch of read-only thread-local Clarity VM database handles (each of which contains a read-only MARF struct) and give each P2P and PRC thread its own MARF.  I'm in favor of the latter approach because it lets us bound the number of file descriptors used by the MARF globally in a very natural way -- it's a function of the (fixed) number of threads we allocate to these other subsystems.  We can also continue to allow `get` to take a mutable self reference.  Trying to accomplish something similar by tracking the number of concurrent `get` calls on a single thread-shared MARF sounds much hairier by comparison.  But we don't have to solve this problem today.